src/main/java/minijava/util/LookAheadIterator.java

```diff
@@ -0,0 +1,94 @@
+package minijava.util;
+
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.Optional;
+
+/** An @Iterator@ decorator which enables look ahead into iterated elements. */
+public class LookAheadIterator<T> implements Iterator<T> {
+
+  private final Iterator<T> it;
+  private final ArrayDeque<T> nextElements = new ArrayDeque<>(16);
+  private boolean isBeforeFirstElement = true;
+
+  public LookAheadIterator(Iterator<T> it) {
+    this.it = it;
+  }
+
+  /**
+   * Looks ahead into the stream of elements, tries to return the element as the result of the n-th
+   * call to @next()@, if there is one, or @Optional.empty()@ otherwise.
+   *
+   * <p>To get a feel for how this works: In the call
+   * sequence @lookAhead(1)@, @next()@, @lookAhead(0)@ on an underlying non-empty iterable, all
+   * calls return the same element (which is the first element of the stream!).
+   *
+   * @lookAhead(0)@ is always the 'current' element, if any. On a non-empty, fresh
+   *     iterable, @lookAhead(0)@ will be @Optional.empty()@, because we start 'before' the first
+   *     element in the stream, respecting @Iterable@ semantics.
+   * @lookAhead(1)@ will point to the first element in that case.
+   *     <p>To fullfill calls to @lookAhead(n)@ with n greater than 1, it will call @next@ on the
+   *     underlying iterable as needed. A call to @lookAhead(0)@ is guaranteed not to call @next@.
+   * @param n non-negative integer, denoting the number of elements to look ahead.
+   * @return The element which will be the return value after @n-1@ calls to @next@.
+   */
+  public Optional<T> lookAhead(int n) {
+    if (n < 0) {
+      throw new IllegalArgumentException("n was negative");

```

Comment:
Since we already have Guava in our project, you could use [checkArgument(n < 0, "n was negative")](https://github.com/google/guava/wiki/PreconditionsExplained) here, which would be slightly more elegant (only a suggestion for the future :wink: )